lib/libzfs/libzfs_crypto.c

```diff
@@ -489,6 +498,126 @@ get_key_material_file(libzfs_handle_t *hdl, const char *uri,
 	return (ret);
 }
 
+static int
+execute_key_fob(libzfs_handle_t *hdl, const char *path,
+    encryption_backend_op_t op, const char *fsname, int *outfd)
+{
+	int ret = 0, spawnret = 0, status = 0, compipe[2];
+	char * const argv[] =
+	    {(char *)path, (char *)backend_ops_lcase[op], (char *)fsname, NULL};
+	pid_t child = 0;
+	posix_spawn_file_actions_t fact;
+
+	if (pipe2(compipe, O_CLOEXEC | O_NONBLOCK) != 0) {
+		ret = errno;
+		errno = 0;
+		zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+		    "Failed to create keyfob pipes"));
+		return (ret);
+	}
+
+	posix_spawn_file_actions_init(&fact);
+	posix_spawn_file_actions_adddup2(&fact, compipe[1], 3);
+
+	spawnret = posix_spawn(&child, argv[0], &fact, NULL, argv, environ);
+	posix_spawn_file_actions_destroy(&fact);
+	close(compipe[1]);
+
+	if (spawnret != 0) {
+		ret = spawnret;
+		zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+		    "Failed to spawn keyfob %s: %s"),
+		    argv[0], strerror(spawnret));
+
+		goto end;
+	}
+
+	if (waitpid(child, &status, 0) < 0) {
+		ret = errno;
+		errno = 0;
+		zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+		    "Failed to wait for keyfob (%d)"), child);
+
+		goto end;
+	}
+
+	if (WIFSIGNALED(status) || WEXITSTATUS(status) != 0) {
+		if (WIFSIGNALED(status)) {
+			ret = EZFS_INTR;
+			zfs_error_aux(hdl, dgettext(TEXT_DOMAIN,
+			    "Keyfob %s killed by %d"),
+			    argv[0], WTERMSIG(status));
```

Comment:
Might be nice to print `strsignal` instead.

Reply:
updated